lib/dry/system/settings.rb

```diff
@@ -0,0 +1,67 @@
+require "dry/core/class_builder"
+require "dry/types"
+require "dry/struct"
+
+require "dry/system/settings/file_loader"
+
+module Dry
+  module System
+    module Settings
+      module Types
+        include Dry::Types.module
+      end
+
+      class DSL < BasicObject
+        attr_reader :identifier
+
+        attr_reader :schema
+
+        def initialize(identifier, &block)
+          @identifier = identifier
+          @schema = {}
+          instance_eval(&block)
+        end
+
+        def call
+          Core::ClassBuilder.new(name: 'Configuration', parent: Settings::Configuration).call do |klass|
+            schema.each do |key, type|
+              klass.setting(key, type)
+            end
+          end
+        end
+
+        def key(name, type)
+          schema[name] = type
+        end
+
+        def type(id)
+          Types::Strict.const_get(id.to_s)
```

Comment:
This is pretty neat, but I've a couple of thoughts:

1. We may want to reconsider whether we make these all strict by default, or if we can add some extra layer of coercion we run things through. When we're getting our settings from the .env files or the actual ENV (e.g. in production), they're all strings. In our apps in the past we've actually used some `Form` types (like `Types::Form::Bool`) to coerce these into proper Ruby values.

2. Is hiding the types behind this `type` method actually going to make things more confusing, i.e. make it not really clear that we're directly using dry-types type objects? e.g. `type(String)`  could easily be mistaken for some special API that expects the `::String` constant to be passed to it. Would it be better to just have the settings refer to e.g. `Types::Strict::String` directly?

Reply:
@timriley I think it's valuable to have them strict by default. I'll add custom types for coercions too (I'm aware that we're using `Form::Bool` in some places).

> Would it be better to just have the settings refer to e.g. Types::Strict::String directly?

Right, I thought about this. I guess I can make `Types::*` work and people can decide which types they want to use.

Reply:
@solnic OK, this sounds like a good plan then (both the support for easy coercions + ability to use types directly). And yeah, I agree that the strictness is a benefit here.